### PR TITLE
Add new `write_board_manifest()` function

### PR DESCRIPTION
**Pull request by juliasilge**
Related to #650 

That PR is a bit sprawling so I think this will be better if we can break it up into some pieces. This PR implements _part_ of that change, adding the new function `board_manifest()` and the internal generic to call for various boards.

I have not implemented this for `board_rsconnect()`, because it's going to be pretty complicated. (Maybe look up all pins that the specific user can access?)

I changed the function name since this operates on a board and returns the board (it doesn't act on a pin).

### CODE REVIEW

**Comment by juliasilge** [comment]
Currently one failure because we can't build the brand-new version of arrow on Windows.

@ijlyttle do you mind taking a look at this PR and seeing if you have feedback on the changes I made relative to your original work?

**Comment by ijlyttle** [comment]
I think this looks great; the naming makes perfect sense!

Sorry I let this get away from me a bit. I agree that reducing the scope of each bit will make it much easier to manage.

Thanks!

**Review by machow — CHANGES_REQUESTED** [review]
Producing a manifest seems super useful! I noticed we are writing a top-level file to a board, so mostly discuss that here, but wonder if it's worth separating out the creating of a manifest file, and this new behavior (a top-level file on a board) into separate PRs?

I'm a bit concerned that writing files (or even worse folders) that aren't pins to a board may produce surprising behaviors.

From a quick test, pins-python will see `pins.txt` as a pin (as will an R pins azure board?). This can be patched pretty quickly, but may take a bit of syncing across libraries / users sharing a board.

For the new behavior of adding a top-level file, can we test the following across all boards?:

* writing a `pins.txt` will not cause it to show up in pin_list
* writing a `pins.txt` will produce the correct result for pin_{exists,meta}("pins.txt")
* writing a `pins.txt` will not cause pins::cache_{info,prune} errors / surprising behavior
* explicate what will happen in pins legacy functions (e.g. could be explicitly unspecified)
* explicate / test what happens if a user attempts to create a pin named `"pins.txt"`
* explicate / test what happens if a user has created a pin named `"pins.txt"` and runs `board_manifest()`

It seems like these could go in `test_api_basic()` or a similarly named function.

Additional notes:

When I try it on s3, I get the warning below -- do you mind double checking?:

image

File: R/board_manifest.R [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    ~fs::path(.x, pin_versions(board, name = .x)$version) %>%
+      append_slash() %>%
+      as.list()
+  )
+  names(result) <- pin_names
+
+  result
+}
+
+append_slash <- function(x) {
```
Comment (ijlyttle) [inline R/board_manifest.R:53]: my bad - I think the warning may be because my helper function `append_slash()` is not vectorized.

Here's a better version:

```r
append_slash <- function(x) {
  has_slash <- grepl("/$", x)
  x[!has_slash] <- paste0(x[!has_slash], "/")
  
  x
}
```

File: R/board_manifest.R [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  if (!grepl("/$", x)) {
+    x <- paste0(x, "/")
+  }
+  x
+}
```
Comment (ijlyttle) [inline R/board_manifest.R:58]: my bad - I think the warning may be because my helper function `append_slash()` is not vectorized.

Here's a better version:

```r
append_slash <- function(x) {
  has_slash <- grepl("/$", x)
  x[!has_slash] <- paste0(x[!has_slash], "/")
  
  x
}
```

File: R/board_manifest.R [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  )
+  names(result) <- pin_names
+
+  result
+}
+
+append_slash <- function(x) {
+  if (!grepl("/$", x)) {
+    x <- paste0(x, "/")
+  }
```
Comment (ijlyttle) [inline R/board_manifest.R:56]: my bad - I think the warning may be because my helper function `append_slash()` is not vectorized.

Here's a better version:

```r
append_slash <- function(x) {
  has_slash <- grepl("/$", x)
  x[!has_slash] <- paste0(x[!has_slash], "/")
  
  x
}
```

Comment (ijlyttle) [reply R/board_manifest.R:56]: also, if it can be useful:

```r
expect_identical(append_slash("foo"), "foo/")
expect_identical(append_slash("foo/"), "foo/")
expect_identical(append_slash(c("foo/", "bar")), c("foo/", "bar/"))
expect_identical(append_slash(c("foo", "bar/")), c("foo/", "bar/"))
```

Comment (juliasilge) [reply R/board_manifest.R]: Thank you! 🙌

**Comment by juliasilge** [comment]
I think I got all the tests and additional documentation added here. 

I don't think I'll open an issue about doing this for Connect until someone asks for it, as it is going to be a bear (just leave it non-implemented so default method gets called). As a reminder, this is the first ~half of what #650 aims for, and there will still be more to do for the overall workflow to be possible.

**Comment by ijlyttle** [comment]
Hi @juliasilge - just a quick note, I have updated my workflows to use this branch, everything is working great with `_pins.yaml` and `write_board_manifest()`. 

This should not be in the least bit surprising, but figured you might like to know.

Thanks!

**Comment by machow** [comment]
Sorry for the wait -- I'm going to double check quickly against pins python boards, but expect everything should work fine!

**Comment by machow** [comment]
@juliasilge it appears adding a `pins.txt` file as a manifest breaks `pin_search` on python pins. The reason is that in the python version `pin_list()` returns the filename, and `pin_search` attempts to retrieve metadata for every pin as part of search.

Here's the relevant R pins [pins_search code fetching metadata](https://github.com/rstudio/pins-r/blob/b0b6a47f654a1b55159991aeb167dbfcbe6ea8a5/R/pin_search.R#L32-L34) (which works fine with the manifest file).

wdyt is a good move here? I can modify pins-python to ignore `pins.txt` files, but I'm wondering what the right way sequence is for releasing..

**Comment by machow** [comment]
> This is helpful in general in case people are putting other files in, say, an S3 bucket that holds pins.

I wonder if we should discourage people from putting other files/folders into a pin board? Ignoring files will work okay, but if people put folders in they'll get listed as pins (afaict the exact behavior varies from board to board).

For example, if a folder with a file in it (`abc/xyz.txt`) is inside of an S3 board, then pin_search will raise an error (because it will try to pull the metadata for a pin named `abc`):

image

**Review by machow — APPROVED** [review]

File: R/board_azure.R [hunk 4]
```diff
@@ -226,13 +245,11 @@ azure_ls <- function(board, dir = "") {
   paths <- AzureStor::list_storage_files(
     board$container,
     dir = dir,
-    recursive = FALSE,
-    info = "name"
+    recursive = FALSE
```
Comment (hadley) [inline R/board_azure.R:248]: Why did this change?

File: R/board_ms365.R [hunk 5]
```diff
@@ -95,11 +95,11 @@ board_ms365 <- function(drive, path, versioned = TRUE, cache = NULL, delete_by_i
 
   cache <- cache %||% board_cache_path(paste0("ms365-", hash(folder$properties$id)))
   new_board_v1("pins_board_ms365",
-    folder = folder,
-    path = path,
-    cache = cache,
-    versioned = versioned,
-    delete_by_item = delete_by_item
+               folder = folder,
```
Comment (hadley) [inline R/board_ms365.R:98]: Unnecessary change?

File: R/board.R [hunk 6]
```diff
@@ -160,6 +160,95 @@ board_deparse.pins_board <- function(board, ...) {
...
+  # given board, return named list:
+  #   - names are pin names
+  #   - values are relative paths to version directories
+
+  pin_names <- pin_list(board)
+
+  result <- map(
+    pin_names,
+    ~fs::path(.x, pin_versions(board, name = .x)$version) %>%
+      append_slash() %>%
```
Comment (hadley) [inline R/board.R:221]: I think it's worth a brief comment explaining why this is needed

File: tests/testthat/test-board_folder.R [hunk 7]
```diff
@@ -43,6 +44,27 @@ test_that("can deparse", {
...
+  write_board_manifest(b)
+
+  # names are correct
+  manifest <- yaml::read_yaml(fs::path(b$path, manifest_pin_yaml_filename))
+  expect_identical(names(manifest), pin_list(b))
+
+  # values (relative paths to versions) are correct
+  imap(
+    manifest,
+    ~ expect_identical(
```
Comment (hadley) [inline tests/testthat/test-board_folder.R:60]: Is there some way you could construct this expected values so the structure is obvious at a glance? And maybe worthwhile to create a pin with more than one version?

File: tests/testthat/test-utils.R [hunk 8]
```diff
@@ -11,3 +11,10 @@ test_that("write_yaml can write non-UTF8 data", {
   y <- yaml::read_yaml(path)
   expect_equal(y, list("fa\u00e7ile" = "fa\u00e7ile"))
 })
+
+test_that("can append slash", {
```
Comment (hadley) [inline tests/testthat/test-utils.R:15]:
```suggestion
test_that("can append slash if needed", {
```

**Review by hadley — APPROVED** [review]

File: R/board_azure.R (hunk 4 above)
Comment (juliasilge) [reply R/board_azure.R:248]: Because we need to know `isdir`, not only the name; see line 250 ⤵️

File: tests/testthat/test-board_folder.R (hunk 7 above)
Comment (juliasilge) [reply tests/testthat/test-board_folder.R:67]: Added another version to one of the test pins 👍 

I thought a while about a way to construct the expected values, but it is not obvious to me with the versions. We currently handle that with `is_testing()` in `version_name()` but that has caused problems as outlined in #640. I wonder if maybe the way to go is snapshot tests with a function to redact versions. I'll add that in the existing issue.
